onnx/defs/printer.cc

```diff
@@ -0,0 +1,154 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "onnx/defs/printer.h"
+
+namespace ONNX_NAMESPACE {
+namespace Utils {
+
+std::ostream& operator<<(std::ostream& os, const TensorShapeProto_Dimension& dim) {
+  if (dim.has_dim_value())
+    os << dim.dim_value();
+  else if (dim.has_dim_param())
+    os << dim.dim_param();
+  else
+    os << "?";
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const TensorShapeProto& shape) {
+  const char* sep = "";
+  const char* comma = ", ";
+  os << "[";
+  for (auto& dim : shape.dim()) {
+    os << sep << dim;
+    sep = comma;
+  }
+  os << "]";
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const TypeProto_Tensor& tensortype) {
+  os << PrimitiveTypeNameMap::ToString(tensortype.elem_type());
+  if (tensortype.has_shape()) {
+    if (tensortype.shape().dim_size() > 0)
+      os << tensortype.shape();
+  } else
+    os << "[...]";
+
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const TypeProto& type) {
+  if (type.has_tensor_type())
+    os << type.tensor_type();
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const ValueInfoProto& value_info) {
+  os << value_info.type() << " " << value_info.name();
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const ValueInfoList& vilist) {
+  const char* sep = "";
+  const char* comma = ", ";
+  os << "(";
+  for (auto& vi : vilist) {
+    os << sep << vi;
+    sep = comma;
+  }
+  os << ")";
+  return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const AttributeProto& attr) {
+  const char* sep = "[";
+  const char* comma = ", ";
+
+  os << attr.name() << " = ";
+  switch (attr.type()) {
+    case AttributeProto_AttributeType_INT:
+      os << attr.i();
+      break;
+    case AttributeProto_AttributeType_INTS:
+      for (auto v : attr.ints()) {
```

Comment:
const auto& ?

Reply:
I think this should be fine for a primitive type like int/float. There is no efficiency gain using a reference to int/float vs using the actual value.

Reply:
For other cases, like strings (or not-yet-handled tensors/graphs), I agree a const-reference should be used.